utils/wasm-builder/src/stack_end.rs

```diff
@@ -99,11 +81,323 @@ pub fn insert_stack_end_export(
+/// If `_start` export exists, but by some reason we cannot insert its call in export functions,
+/// then returns Error.
+pub fn insert_start_call_in_export_funcs(module: &mut Module) -> Result<(), &'static str> {
+    let start_func_index = if let Some(start) = module
+        .export_section()
+        .ok_or("Cannot find export section")?
+        .entries()
+        .iter()
+        .find(|export| export.field() == "_start")
+    {
+        match start.internal() {
+            Internal::Function(index) => *index,
+            _ => return Err("_start export is not a function"),
+        }
+    } else {
+        return Ok(());
+    };
+
+    for export_name in optimize::FUNC_EXPORTS {
+        let Some(export) = module
+            .export_section()
+            .ok_or("Cannot find export section")?
+            .entries()
+            .iter()
+            .find(|export| export.field() == export_name) else
+        {
+            continue
+        };
+
+        let index = match export.internal() {
+            Internal::Function(index) => *index,
+            _ => return Err("Func export is not a function"),
+        };
+
+        let index_in_functions = (index as usize)
+            .checked_sub(module.import_count(ImportCountType::Function))
+            .ok_or("Cannot process case when export function is import")?;
+
+        module.code_section_mut().unwrap().bodies_mut()[index_in_functions]
+            .code_mut()
+            .elements_mut()
+            .insert(0, Instruction::Call(start_func_index));
+    }
+
+    Ok(())
+}
+
+/// For each mutable global, except stack pointer, creates buffer in memory and
+/// initial constant value in data section.
+/// For correct work required stack pointer and data end global names to be
+/// in custom names section.
+///
+/// Returns error if cannot move globals to static memory by some reasons.
+pub fn move_mut_globals_to_static(module: &mut Module) -> Result<(), &'static str> {
+    let module_bytes = module
+        .clone()
+        .to_bytes()
+        .map_err(|_| "cannot get code from module")?;
+
+    // Identify stack pointer and data end globals
+    let stack_pointer_index =
+        get_global_index(&module_bytes, |name| name.ends_with("__stack_pointer"))
+            .ok_or("Cannot find stack pointer global")?;
+    let data_end_index = get_global_index(&module_bytes, |name| name.ends_with("__data_end"))
+        .ok_or("Cannot find data end global")?;
+
+    // Identify mutable globals and their initial data
+    let mut mut_globals = vec![];
+    for (index, global) in module
+        .global_section()
+        .ok_or("Cannot find globals section")?
+        .entries()
+        .iter()
+        .enumerate()
+    {
+        if !global.global_type().is_mutable() {
+            continue;
+        }
+        if index == data_end_index as usize {
+            continue;
+        }
+        if index == stack_pointer_index as usize {
+            continue;
+        }
```

Comment:
Why not `if condition1 || condition2 || condition3 { continue }`.

Reply:
DOne